src/layout.cpp

```diff
@@ -0,0 +1,476 @@
+    }
+  }
+}
+
+void heapify(std::vector<TreeNode> &nodes, int n, int i) {
+  int largest = i;
+  int l = 2 * i + 1;
+  int r = 2 * i + 2;
+
+  if (l < n && nodes[l].id > nodes[largest].id) {
+    largest = l;
+  }
+
+  if (r < n && nodes[r].id > nodes[largest].id) {
+    largest = r;
+  }
+
+  if (largest != i) {
+    std::swap(nodes[i], nodes[largest]);
+    heapify(nodes, n, largest);
+  }
+}
+
+void idHeapSort(std::vector<TreeNode> &nodes) {
+  int n = 0;
+  for (int i = 0; i < nodes.size(); i++) {
+    n++;
+  }
+
+  for (int i = n / 2 - 1; i >= 0; i--) {
+    heapify(nodes, n, i);
+  }
+
+  for (int i = n - 1; i >= 0; i--) {
+    std::swap(nodes[0], nodes[i]);
+    heapify(nodes, i, 0);
+  }
+}
+
+bool algorithmDFS(TreeNode::nodeId id, std::vector<TreeNode> &nodes,
+                  std::vector<bool> &usedId) {
+  usedId[id] = true;
+
+  for (int i = 0; i < nodes[id].succ.size(); i++) {
+    if (usedId[nodes[id].succ[i]]) {
+      return true;
+    } else {
+      bool exist = algorithmDFS(nodes[id].succ[i], nodes, usedId);
+      if (exist) {
+        return true;
+      }
+    }
+  }
+
+  usedId[id] = false;
+  return false;
+}
+
+bool cycleExistsDFS(std::vector<TreeNode> &nodes) {
+  degReinit(nodes);
+
+  std::vector<bool> usedId = {};
+  usedId.resize(nodes.size(), false);
+
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].degM == 0) {
+      bool exist = algorithmDFS(nodes[i].id, nodes, usedId);
+      if (exist) {
+        return true;
+      }
+    }
+  }
+
+  return algorithmDFS(nodes[0].id, nodes, usedId);
+}
+
+struct EdgeCount {
+  int degP = 0;
+  int degM = 0;
+};
+
+void removeEdgeCount(int i, std::vector<EdgeCount> &edgeCounts,
+                     std::vector<TreeNode> &nodes) {
+  for (int j = 0; j < nodes[i].succ.size(); j++) {
+    edgeCounts[nodes[i].succ[j]].degM--;
+  }
+  for (int j = 0; j < nodes[i].pred.size(); j++) {
+    edgeCounts[nodes[i].pred[j]].degP--;
+  }
+
+  edgeCounts[i].degM = -1;
+  edgeCounts[i].degP = -1;
+}
+
+void greedyFAS(
+    std::vector<TreeNode> &nodes,
+    std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges) {
+  degReinit(nodes);
+  predReinit(nodes);
+  std::vector<TreeNode::nodeId> s1 = {}, s2 = {};
+  std::vector<EdgeCount> edgeCounts = {};
+  for (int i = 0; i < nodes.size(); i++) {
+    EdgeCount edgeCount = {};
+    edgeCount.degP = nodes[i].degP;
+    edgeCount.degM = nodes[i].degM;
+    edgeCounts.push_back(edgeCount);
+  }
+  int lenNodes = static_cast<int>(nodes.size());
+  while (lenNodes > 0) {
+    for (int i = 0; i < edgeCounts.size(); i++) {
+      if (edgeCounts[i].degP == 0) {
+        s2.insert(s2.cbegin(), i);
+        removeEdgeCount(i, edgeCounts, nodes);
+        lenNodes--;
+
+        i = -1;
+      }
+    }
+
+    for (int i = 0; i < edgeCounts.size(); i++) {
+      if (edgeCounts[i].degM == 0) {
+        s1.push_back(i);
+        removeEdgeCount(i, edgeCounts, nodes);
+        lenNodes--;
+
+        i = -1;
+      }
+    }
```

Comment:
`i = -1` is an actually a bad break. Here are some observations:
- some sources and sinks are known from the start, so we can find and remember them
- additional sources and sinks are generated *only* by edge-removal, i.e. `removeEdgeCount`, so we can report new sources and sinks and remember them

The observations allows us to use some sort of queue (e.g. `std::vector`-based) and proceed until the queue is empty, e.g.
```
std::vector<TreeNode::nodeId> queue = findSourcesAndSinks(nodes, edgeCounts);
while (!queue.empty()) {
  NodeId id = queue.back();
  queue.pop_back();
  
  if (isSource(id)) {
    s1.push_back(id);
  } else {
    s2.push_back(id);
  }
  removeEdgesAndDetectSourcesAndSinks(nodes[id], nodes, edgeCounts, queue);
}
```